ddl/dm_schema_tracker.go

```diff
@@ -0,0 +1,155 @@
+// limitations under the License.
+
+package ddl
+
+import (
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/table/tables"
+)
+
+// infoStore is a simple structure that stores DBInfo and TableInfo. It's not thread-safe.
+type infoStore struct {
+	lowerCaseTableNames int // same as variable lower_case_table_names
+
+	dbs    map[string]*model.DBInfo
+	tables map[string]map[string]*model.TableInfo
+}
+
+func newInfoStore(lowerCaseTableNames int) *infoStore {
+	return &infoStore{
+		lowerCaseTableNames: lowerCaseTableNames,
+		dbs:                 map[string]*model.DBInfo{},
+		tables:              map[string]map[string]*model.TableInfo{},
+	}
+}
+
+func (i *infoStore) ciStr2Key(name model.CIStr) string {
+	if i.lowerCaseTableNames == 0 {
+		return name.O
+	}
+	return name.L
+}
+
+// SchemaByName is exported to be used when infoStore is embedded into another public struct.
+func (i *infoStore) SchemaByName(name model.CIStr) *model.DBInfo {
+	key := i.ciStr2Key(name)
+	return i.dbs[key]
+}
+
+func (i *infoStore) putSchema(dbInfo *model.DBInfo) {
+	key := i.ciStr2Key(dbInfo.Name)
+	i.dbs[key] = dbInfo
+	if i.tables[key] == nil {
+		i.tables[key] = map[string]*model.TableInfo{}
+	}
+}
+
+func (i *infoStore) deleteSchema(name model.CIStr) bool {
+	key := i.ciStr2Key(name)
+	_, ok := i.dbs[key]
+	if !ok {
+		return false
+	}
+	delete(i.dbs, key)
+	delete(i.tables, key)
+	return true
+}
+
+// TableByName is exported to be used when infoStore is embedded into another public struct.
+func (i *infoStore) TableByName(schema, table model.CIStr) (*model.TableInfo, error) {
+	schemaKey := i.ciStr2Key(schema)
+	tables, ok := i.tables[schemaKey]
+	if !ok {
+		return nil, infoschema.ErrDatabaseNotExists.GenWithStackByArgs(schema)
+	}
+
+	tableKey := i.ciStr2Key(table)
+	tbl, ok := tables[tableKey]
+	if !ok {
+		return nil, infoschema.ErrTableNotExists.GenWithStackByArgs(schema, table)
+	}
+	return tbl, nil
+}
+
+func (i *infoStore) putTable(schemaName model.CIStr, tblInfo *model.TableInfo) error {
+	schemaKey := i.ciStr2Key(schemaName)
+	tables, ok := i.tables[schemaKey]
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(schemaName)
+	}
+	tableKey := i.ciStr2Key(tblInfo.Name)
+	tables[tableKey] = tblInfo
+	return nil
+}
+
+func (i *infoStore) deleteTable(schema, table model.CIStr) error {
+	schemaKey := i.ciStr2Key(schema)
+	tables, ok := i.tables[schemaKey]
+	if !ok {
+		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(schema)
+	}
+
+	tableKey := i.ciStr2Key(table)
+	_, ok = tables[tableKey]
+	if !ok {
+		return infoschema.ErrTableNotExists.GenWithStackByArgs(schema, table)
+	}
+	delete(tables, tableKey)
+	return nil
+}
+
+// infoSchemaAdaptor convert infoStore to infoschema.InfoSchema, it only implements a part of InfoSchema interface to be
+// used by DDL interface.
+// nolint:unused
```

Comment:
Could we delete `nolint:unused`?

Reply:
it will cause lint error. You mean delete infoSchemaAdaptor in this PR and introduce it next PR?

Reply:
Got it.